agent/util-scripts/pbench-tool-meister-start

```diff
@@ -1210,6 +1249,39 @@ def main(_prog: str, cli_params: Namespace) -> int:
                 ReturnCode.TDSWAITFAILURE, "TDS didn't confirm init sequence completion"
             )
 
+        # +
+        # Step 8. - Verify all the Tool Meisters have reported back, and that
+        #           their tool install checks all passed.
+        # -
+        try:
+            tms = json.loads(redis_client.get(tm_data_key))
+        except Exception as exc:
+            error_log(f"Error loading operational Tool Meister data, '{exc}'")
+            raise CleanupTime(
+                ReturnCode.INVALIDTMDATA,
+                "Failed to load reported Tool Meister operational data",
+            )
+
+        tool_install_failures = {}
+        for host, tm in tms.items():
+            if not tm["failed_tools"]:
+                continue
+            tool_install_failures[host] = tm
+
+        if tool_install_failures:
+            error_log("Tool installation checks failed")
+            for host, tm in tool_install_failures.items():
+                for tool_name, (failure_code, output) in tm["installs"].items():
+                    if failure_code != 0:
+                        error_log(
+                            f"{host}: {tool_name} return code: {failure_code},"
+                            f" output: '{output}'"
+                        )
+            raise CleanupTime(
+                ReturnCode.TOOLINSTALLFAILURES,
+                "Tool installation check failures encountered",
+            )
```

Comment:
It would be nice if this blob of code were moved to one or more subroutines, so that it would be reasonable to unit test it.

It would be good if lines 1274-1279 were moved to a subroutine which could be called at line 1269.  This would free us from having to build the `tool_install_failures` mapping (and remove the need to traverse it), allowing us to reduce it to a simple boolean.

Reply:
Sure :)